File: examples/proxy-nodejs/proxy-example.js
/**
 * Route HTTP requests through an Archonum proxy (Node.js + fetch/undici).
 *
 * Archonum hands out per-country HTTP proxies. We ask the API for one, then
 * make a normal fetch through it — the exit IP comes out in the chosen
 * country. Node's global fetch doesn't speak proxies, so we route it through
 * undici's ProxyAgent (undici is Node's own HTTP client, added as a dependency
 * here so we can import ProxyAgent).
 *
 *   GET {ARCHONUM_BASE_URL}/api/v1/proxies/?country=..&session=..
 *     Authorization: Token {ARCHONUM_TOKEN}
 *   -> { host, http_port, username, password, country, session }
 *
 * Needs: ARCHONUM_TOKEN, ARCHONUM_COUNTRY
 */

import { ProxyAgent } from 'undici';

const DEFAULT_BASE_URL = 'https://app.archonum.com';
const TARGET_URL = process.env.TARGET_URL || 'https://api.ipify.org?format=json';

/** Ask the Archonum API for a per-country HTTP proxy. */
async function fetchProxy() {
  const token = process.env.ARCHONUM_TOKEN;
  const country = (process.env.ARCHONUM_COUNTRY || '').toLowerCase();
  if (!token || !country) throw new Error('set ARCHONUM_TOKEN and ARCHONUM_COUNTRY');
  const base = (process.env.ARCHONUM_BASE_URL || DEFAULT_BASE_URL).replace(/\/$/, '');
  const session = process.env.ARCHONUM_SESSION || 'example';

  const url = `${base}/api/v1/proxies/?country=${country}&session=${session}`;
  console.log(`fetching Archonum proxy -> ${url}`);
  const res = await fetch(url, { headers: { Authorization: `Token ${token}` } });
  if (!res.ok) throw new Error(`Archonum returned HTTP ${res.status}`);
  const body = await res.json();
  if (body.error) throw new Error(`Archonum returned no proxy for ${country}: ${body.error}`);
  console.log(`  proxy ${body.host}:${body.http_port} country=${body.country}`);
  return body;
}

async function main() {
  const proxy = await fetchProxy();
  // uri is the proxy address; token is the Proxy-Authorization header value.
  const dispatcher = new ProxyAgent({
    uri: `http://${proxy.host}:${proxy.http_port}`,
    token: `Basic ${Buffer.from(`${proxy.username}:${proxy.password}`).toString('base64')}`,
  });

  // Node's global fetch has no proxy option, so route it through the dispatcher.
  console.log(`requesting ${TARGET_URL} through the proxy...`);
  const res = await fetch(TARGET_URL, { dispatcher });
  const text = (await res.text()).trim();
  console.log(`  status -> ${res.status}`);
  console.log(`  body   -> ${text}`);
}

main().catch((e) => {
  console.error(e.message || e);
  process.exit(1);
});

File: examples/proxy-python/proxy_example.py
"""
Route HTTP requests through an Archonum proxy (Python + requests).

Archonum hands out per-country HTTP proxies. We ask the API for one, then make
a normal request through it — the exit IP comes out in the chosen country.

  GET {ARCHONUM_BASE_URL}/api/v1/proxies/?country=..&session=..
    Authorization: Token {ARCHONUM_TOKEN}
  -> { host, http_port, username, password, country, session }

Needs: ARCHONUM_TOKEN, ARCHONUM_COUNTRY
"""

import json
import os

import requests

DEFAULT_BASE_URL = "https://app.archonum.com"
TARGET_URL = os.environ.get("TARGET_URL", "https://api.ipify.org?format=json")
HTTP_TIMEOUT = float(os.environ.get("HTTP_TIMEOUT_MS", "15000")) / 1000


def fetch_proxy() -> dict:
    """Ask the Archonum API for a per-country HTTP proxy."""
    token = os.environ.get("ARCHONUM_TOKEN")
    country = (os.environ.get("ARCHONUM_COUNTRY") or "").lower()
    if not token or not country:
        raise SystemExit("set ARCHONUM_TOKEN and ARCHONUM_COUNTRY")
    base = os.environ.get("ARCHONUM_BASE_URL", DEFAULT_BASE_URL).rstrip("/")
    session = os.environ.get("ARCHONUM_SESSION", "example")

    url = f"{base}/api/v1/proxies/?country={country}&session={session}"
    print(f"fetching Archonum proxy -> {url}")
    res = requests.get(url, headers={"Authorization": f"Token {token}"}, timeout=HTTP_TIMEOUT)
    res.raise_for_status()
    body = res.json()
    if "error" in body:
        raise SystemExit(f"Archonum returned no proxy for {country}: {body['error']}")
    print(f"  proxy {body['host']}:{body['http_port']} country={body['country']}")
    return body


def proxy_url(proxy: dict) -> str:
    return f"http://{proxy['username']}:{proxy['password']}@{proxy['host']}:{proxy['http_port']}"


def main() -> None:
    proxy = fetch_proxy()
    proxies = {"http": proxy_url(proxy), "https": proxy_url(proxy)}

    print(f"requesting {TARGET_URL} through the proxy...")
    res = requests.get(TARGET_URL, proxies=proxies, timeout=HTTP_TIMEOUT)
    res.raise_for_status()
    print(f"  status -> {res.status_code}")
    print(f"  body   -> {res.text.strip()}")


if __name__ == "__main__":
    main()
